### PR TITLE
solana-ibc: use solana height and timestamp for cf-guest host metadata

### PR DESCRIPTION
**Pull request by dhruvja**
since the cf-guest light client would exist on rollup, the host metadata is the solana slot and the timestamp instead of those fetched from the guest chain since the rollup wouldnt have it.

But if the connection is established to another svm where the guest chain is used, then the above should be reverted.

### CODE REVIEW

File: solana/solana-ibc/programs/solana-ibc/src/client_state.rs [hunk 1]
```diff
@@ -236,14 +237,20 @@ impl cf_guest::CommonContext<sigverify::ed25519::PubKey>
...
+                description: e.to_string(),
             })?;
 
-        let height = u64::from(self.borrow().chain.head()?.block_height);
-        let height = ibc::Height::new(1, height)?;
+        let slot = clock.slot;
+        let timestamp_sec = clock.unix_timestamp as u64;
+
+        let timestamp =
+            ibc::Timestamp::from_nanoseconds(timestamp_sec * 10u64.pow(9))
```
Comment (mina86) [inline solana/solana-ibc/programs/solana-ibc/src/client_state.rs:249]:
```suggestion
            ibc::Timestamp::from_nanoseconds(timestamp_sec * 10u64.ipow(9))
```

**Comment by mina86** [comment]
I think this needs to be conditional. If we’re running on Solana we have chain account and we use time from there. But if witness account is enabled, we don’t have chain account and then we read data directly from blockchain.

File: solana/solana-ibc/programs/solana-ibc/src/client_state.rs (hunk 1 above)
Comment (dhruvja) [reply solana/solana-ibc/programs/solana-ibc/src/client_state.rs]: we would need `num` dependency if we wanna use `ipow` right?

Comment (mina86) [reply solana/solana-ibc/programs/solana-ibc/src/client_state.rs]: Sorry, ignore this.

File: solana/solana-ibc/programs/solana-ibc/src/validation_context.rs [hunk 2]
```diff
@@ -43,12 +49,41 @@ impl ibc::ValidationContext for IbcStorage<'_, '_> {
...
+                        description: e.to_string(),
+                    })?;
+
+            let slot = clock.slot;
+            let height = ibc::Height::new(1, slot)?;
+            return Ok(height);
+        }
         let height = u64::from(self.borrow().chain.head()?.block_height);
         let height = ibc::Height::new(1, height)?;
         Ok(height)
```
Comment (mina86) [inline solana/solana-ibc/programs/solana-ibc/src/validation_context.rs:66]: In general, prefer `cfg!` macro to conditional compilation:

```suggestion
        let height = if cfg!(feature = "witness") {
            anchor_lang::solana_program::sysvar::clock::Clock::get()
                .map_err(|e| ibc::ClientError::ClientSpecific {
                    description: e.to_string(),
                })?.slot
        } else {
            u64::from(self.borrow().chain.head()?.block_height)
        };
        ibc::Height::new(1, height)
```

It often makes the code easier to read and makes build failure more
obvious.  With conditional compilation, you need to build with all
combination of features to make sure everything compiles correctly.
With `cfg!` macro, everything is always well formed.

File: solana/solana-ibc/programs/solana-ibc/src/validation_context.rs [hunk 3]
```diff
@@ -43,12 +49,41 @@ impl ibc::ValidationContext for IbcStorage<'_, '_> {
...
+                timestamp_sec * 10u64.pow(9),
+            )
+            .map_err(|e| {
+                ibc::ClientError::ClientSpecific { description: e.to_string() }
+                    .into()
+            });
+        }
         let timestamp = self.borrow().chain.head()?.timestamp_ns.get();
         ibc::Timestamp::from_nanoseconds(timestamp).map_err(|err| {
             ibc::ClientError::Other { description: err.to_string() }.into()
```
Comment (mina86) [inline solana/solana-ibc/programs/solana-ibc/src/validation_context.rs:89]:
```suggestion
        let timestamp_ns = if cfg!(feature = "witness") {
            anchor_lang::solana_program::sysvar::clock::Clock::get()
                .map_err(|e| ibc::ClientError::ClientSpecific {
                    description: e.to_string(),
                })?.clock.unix_timestamp as u64 * 10u64.pow(9)
        } else {
            self.borrow().chain.head()?.timestamp_ns.get()
        };
        ibc::Timestamp::from_nanoseconds(timestamp).map_err(|err| {
            ibc::ClientError::ClientSpecific { description: err.to_string() }.into()
```

File: solana/solana-ibc/programs/solana-ibc/src/validation_context.rs [hunk 4]
```diff
@@ -2,14 +2,15 @@ use std::str::FromStr;
...
+#[cfg(feature = "witness")]
+use anchor_lang::prelude::SolanaSysvar;
 use lib::hash::CryptoHash;
 
 use crate::client_state::AnyClientState;
 use crate::consensus_state::AnyConsensusState;
 use crate::ibc::{self, ConsensusState};
 use crate::storage::{self, IbcStorage};
 
-
```
Comment (mina86) [inline solana/solana-ibc/programs/solana-ibc/src/validation_context.rs:12]: bit of diff noise.

**Review by mina86 — APPROVED** [review]

File: solana/solana-ibc/programs/solana-ibc/src/validation_context.rs (hunk 2 above)
Comment (dhruvja) [reply solana/solana-ibc/programs/solana-ibc/src/validation_context.rs:66]: i see, that makes sense

File: solana/solana-ibc/programs/solana-ibc/src/validation_context.rs (hunk 4 above)
Comment (dhruvja) [reply solana/solana-ibc/programs/solana-ibc/src/validation_context.rs:12]: gotcha

**Pull request merged**